Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,780 @@
+			ignoreList.Add(go);
+
+			for (int i = 0; i < moduleIgnoreList.Length; i++)
+			{
+				ignoreList.Add(moduleIgnoreList[i].gameObject);
+			}
+		}
+
+		static void SetupIgnoreList(List<GameObject> ignoreList, Renderer[] moduleIgnoreList, GameObject[] objects)
+		{
+			ignoreList.Clear();
+
+			for (int i = 0; i < objects.Length; i++)
+			{
+				ignoreList.Add(objects[i]);
+			}
+
+			for (int i = 0; i < moduleIgnoreList.Length; i++)
+			{
+				ignoreList.Add(moduleIgnoreList[i].gameObject);
+			}
+		}
+
+		bool SnapToSurface(Ray ray, ref Vector3 position, ref Quaternion rotation, SnappingState state, Vector3 boundsOffset, Vector3 targetPosition, Quaternion targetRotation, Quaternion rotationOffset, Vector3 upVector, List<GameObject> ignoreList, float breakDistance, float raycastDistance)
+		{
+			RaycastHit hit;
+			GameObject go;
+			if (raycast(ray, out hit, out go, raycastDistance, ignoreList))
+			{
+				var snappedRotation = Quaternion.LookRotation(hit.normal, upVector) * rotationOffset;
+
+				var hitPoint = hit.point;
+				m_CurrentSurfaceSnappingHit = hitPoint;
+				var snappedPosition = pivotSnapping ? hitPoint : hitPoint + rotation * boundsOffset;
+
+				if (localOnly && Vector3.Distance(snappedPosition, targetPosition) > breakDistance)
+					return false;
+
+				state.surfaceSnapping = true;
+				state.groundSnapping = false;
+
+				position = snappedPosition;
+				rotation = snapRotation ? snappedRotation : targetRotation;
+
+				m_CurrentSurfaceSnappingPosition = position;
+				m_CurrentSurfaceSnappingRotation = snappedRotation;
+				return true;
+			}
+
+			return false;
+		}
+
+
+		bool SnapToGround(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation, SnappingState state, float groundSnapMin, float groundSnapMax)
+		{
+			if (groundSnappingEnabled)
+			{
+				var diffGround = Mathf.Abs(targetPosition.y - k_GroundHeight);
+
+				var bounds = state.rotatedBounds;
+				if (snapRotation)
+					bounds = state.identityBounds;
+
+				var offset = bounds.center.y - bounds.extents.y;
+
+				if (!pivotSnapping)
+					diffGround = Mathf.Abs(targetPosition.y + offset - k_GroundHeight);
+
+				if (diffGround < groundSnapMin)
+					state.groundSnapping = true;
+
+				if (diffGround > groundSnapMax)
+				{
+					state.groundSnapping = false;
+					position = targetPosition;
+					rotation = targetRotation;
+				}
+
+				if (state.groundSnapping)
+				{
+					if (pivotSnapping)
+						targetPosition.y = k_GroundHeight;
+					else
+						targetPosition.y = k_GroundHeight - offset;
+
+					position = targetPosition;
+
+					if (snapRotation)
+						rotation = Quaternion.LookRotation(Vector3.up, targetRotation * Vector3.back) * Quaternion.AngleAxis(90, Vector3.right);
+
+					return true;
+				}
+			}
+
+			return false;
+		}
+
+		SnappingState GetSnappingState(Transform rayOrigin, GameObject go, Vector3 position, Quaternion rotation)
+		{
+			Dictionary<GameObject, SnappingState> states;
```

Comment:
Instead of copy/pasted code, can you keep a cache field for a local array of 1 element that you can set and use the other method below?